source/uri/Locator.ooc

```diff
@@ -0,0 +1,99 @@
+use uri
+use collections
+
+Locator: class {
+	_scheme: Text
+	_authority: Authority
+	_path: VectorList<Text>
+	_query: Query
+	_fragment: Text
+	scheme: Text {
+		get { this _scheme take() }
+	}
+	authority: Authority {
+		get { this _authority }
+	}
+	path: VectorList<Text> {
+		get {
+			result: VectorList<Text>
+			if (this _path != null) {
+				result = VectorList<Text> new()
+				for (i in 0 .. this _path count) {
+					result add(this _path[i] take())
+				}
+			}
+			result
+		}
+	}
+	query: Query {
+		get { this _query }
+	}
+	fragment: Text {
+		get { this _fragment take() }
+	}
+	init: func(=_scheme, =_authority, =_path, =_query, =_fragment)
+	free: override func {
+		this _scheme free(Owner Receiver)
+		if (this _authority != null)
+			this _authority free()
+		if (this _path != null) {
+			for (i in 0 .. this _path count)
+				this _path[i] free(Owner Receiver)
+			this _path free()
+		}
+		if (this _query != null)
+			this _query free()
+		this _fragment free(Owner Receiver)
+		super()
+	}
+	toText: func -> Text {
+		result := Text new()
+		if (!this scheme isEmpty)
+			result += this scheme + t"://"
+		if (this authority != null)
+			result += this authority toText()
+		path := this path
+		if (path != null) {
+			for (i in 0 .. path count)
+				result += t"/" + path[i]
+			path free()
+		}
+		if (this query != null)
+			result += t"?" + this query toText()
+		if (!this fragment isEmpty)
+			result += t"#" + this fragment
+		result
+	}
+	parse: static func(data: Text) -> This {
+		result: This
+		d := data take()
+		if (!d isEmpty) {
+			splitted := d split(t"://")
+			scheme := splitted count > 1 ? splitted remove(0) : Text empty
+			d = splitted remove()
```

Comment:
We generally don't do that level of defensive programming. Maybe in a version block.